bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/EventFilterHandler.java

```diff
@@ -0,0 +1,362 @@
+                    return;
+                }
+            }
+
+            Instant reference = Instant.now();
+            TimeMultiplicator multiplicator;
+
+            try {
+                if (config.datetimeUnit == null) {
+                    throw new IllegalArgumentException("Config is broken.");
+                }
+                multiplicator = TimeMultiplicator.valueOf(config.datetimeUnit);
+            } catch (IllegalArgumentException e) {
+                logger.warn("Time-filter settings are not set properly.");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+                return;
+            }
+
+            if (config.datetimeRound) {
+                ZonedDateTime refDT = reference.atZone(ZoneId.systemDefault());
+                switch (multiplicator) {
+                    case WEEK:
+                        refDT = refDT.with(ChronoField.DAY_OF_WEEK, 1);
+                    case DAY:
+                        refDT = refDT.with(ChronoField.HOUR_OF_DAY, 0);
+                    case HOUR:
+                        refDT = refDT.with(ChronoField.MINUTE_OF_HOUR, 0);
+                    case MINUTE:
+                        refDT = refDT.with(ChronoField.SECOND_OF_MINUTE, 0);
+                }
+                reference = refDT.toInstant();
+            }
+            Instant begin = Instant.EPOCH;
+            Instant end = Instant.ofEpochMilli(Long.MAX_VALUE);
+
+            if (config.datetimeStart != null) {
+                begin = reference.plusSeconds(config.datetimeStart.longValue() * multiplicator.getMultiplier());
+            }
+            if (config.datetimeEnd != null) {
+                end = reference.plusSeconds(config.datetimeEnd.longValue() * multiplicator.getMultiplier());
+            }
+
+            List<Event> results = cal.getFilteredEventsBetween(begin, end, filter, config.maxEvents.intValue());
+            for (int position = 0; position < config.maxEvents.intValue(); position++) {
+                ResultChannelSet channels = resultChannels.get(position);
+                if (position < results.size()) {
+                    Event result = results.get(position);
+                    updateState(channels.titleChannel, new StringType(result.title));
+                    updateState(channels.beginChannel, new DateTimeType(result.start.atZone(ZoneId.systemDefault())));
+                    updateState(channels.endChannel, new DateTimeType(result.end.atZone(ZoneId.systemDefault())));
+                } else {
+                    updateState(channels.titleChannel, UnDefType.UNDEF);
+                    updateState(channels.beginChannel, UnDefType.UNDEF);
+                    updateState(channels.endChannel, UnDefType.UNDEF);
+                }
+            }
+
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```

Comment:
Please add an appropriate error reason to give the user a hint when debugging.

Reply:
OK